src/prommis/nanofiltration/nf_brine_plot.py

```diff
@@ -0,0 +1,498 @@
+    """
+    Fixes the initial variables needed to create 0 DOF
+    """
+
+    # pump variables
+    m.fs.pump.efficiency_pump[0].fix(0.75)
+    m.fs.pump.control_volume.properties_in[0].temperature.fix(298.15)
+    m.fs.pump.control_volume.properties_in[0].pressure.fix(101325)
+    m.fs.pump.outlet.pressure[0].fix(2e5)
+    iscale.set_scaling_factor(m.fs.pump.control_volume.work, 1e-4)
+
+    # membrane operation
+    m.fs.unit.recovery_vol_phase[0, "Liq"].setub(0.95)
+    m.fs.unit.spacer_porosity.fix(0.85)
+    m.fs.unit.channel_height.fix(5e-4)
+    m.fs.unit.velocity[0, 0].fix(0.1)
+    m.fs.unit.area.fix(100)
+    m.fs.unit.mixed_permeate[0].pressure.fix(101325)
+
+    # variables for calculating mass transfer coefficient with spiral wound correlation
+    m.fs.unit.spacer_mixing_efficiency.fix()
+    m.fs.unit.spacer_mixing_length.fix()
+
+    # membrane properties
+    m.fs.unit.radius_pore.fix(0.5e-9)
+    m.fs.unit.membrane_thickness_effective.fix(1.33e-6)
+    m.fs.unit.membrane_charge_density.fix(-60)
+    m.fs.unit.dielectric_constant_pore.fix(41.3)
+    iscale.calculate_scaling_factors(m)
+
+
+def unfix_opt_vars(m):
+    """
+    Unfixes select variables to enable optimization with DOF>0
+    """
+    m.fs.pump.outlet.pressure[0].unfix()
+    m.fs.unit.area.unfix()
+
+
+def add_obj(m):
+    """
+    Adds objective to the pyomo model
+    """
+    # limit Li loss
+    m.fs.obj = Objective(
+        expr=m.fs.retentate.flow_mol_phase_comp[0, "Liq", "Li_+"], sense=minimize
+    )
+
+    # # maxmize the reduction in Mg:Li ratio
+    # m.fs.mg_li_ratio_obj = Objective(
+    #     expr=(
+    #         (
+    #             (m.fs.feed.flow_mol_phase_comp[0, "Liq", "Mg_2+"].value / 0.024)
+    #             / (m.fs.feed.flow_mol_phase_comp[0, "Liq", "Li_+"].value / 0.0069)
+    #         )
+    #         - (
+    #             (m.fs.permeate.flow_mol_phase_comp[0, "Liq", "Mg_2+"].value / 0.024)
+    #             / (m.fs.permeate.flow_mol_phase_comp[0, "Liq", "Li_+"].value / 0.0069)
+    #         )
+    #     ),
+    #     sense=maximize
+    # )
+
+
+def add_pressure_con(m, pressure_limit):
+    """
+    Adds feed pressure constraint to the pyomo model
+    """
+    if pressure_limit is None:
+        pressure_limit = 7e6
+    # bound the feed pressure to a reasonable value for nanofiltration
+    # choose an upper limit of 70 bar (https://doi.org/10.1021/acs.est.2c08584)
+    m.fs.pressure_con = Constraint(expr=m.fs.pump.outlet.pressure[0] <= pressure_limit)
+
+
+def add_recovery_con(m, recovery_limit):
+    """
+    Adds recovery constraint to the pyomo model
+    """
+    if recovery_limit is None:
+        recovery_limit = 0.8
+    # limit the NF recovery
+    m.fs.recovery_con = Constraint(
+        expr=m.fs.unit.recovery_vol_phase[0.0, "Liq"] <= recovery_limit
+    )
+
+
+def optimize(m, solver):
+    """
+    Optimizes the flowsheet
+    """
+    print(f"Optimizing with {format(degrees_of_freedom(m))} DOFs")
```

Comment:
Again, I suggest using a logger and raising an exception rather than an assertion.